src/uglify/index.js

```diff
@@ -14,77 +13,69 @@ try {
 } catch (e) { } // eslint-disable-line no-empty
 
 export default class {
-  constructor(parallel = {}) {
-    let options = parallel;
-    if (typeof parallel === 'boolean') {
-      options = { cache: parallel, workers: parallel };
-    }
-    const { cache, workers } = options;
-    this.cache = cache === true ? findCacheDir({ name: 'uglifyjs-webpack-plugin' }) : cache;
-    this.workers = workers === true ? os.cpus().length - 1 : Math.min(Number(workers) || 0, os.cpus().length - 1);
+  constructor(options = {}) {
+    const { cache, parallel } = options;
+    this.cacheDir = cache === true ? findCacheDir({ name: 'uglifyjs-webpack-plugin' }) : cache;
+    this.maxConcurrentWorkers = parallel === true ? os.cpus().length - 1 : Math.min(Number(parallel) || 0, os.cpus().length - 1);
   }
 
-  worker(options, callback) {
-    if (this.workers > 0) {
-      this.workerFarm = workerFarm({
-        maxConcurrentWorkers: this.workers,
+  runTasks(tasks, callback) {
+    if (!tasks.length) {
+      callback(null, []);
+      return;
+    }
+
+    if (this.maxConcurrentWorkers > 0) {
+      this.workers = workerFarm({
+        maxConcurrentWorkers: this.maxConcurrentWorkers,
       }, workerFile);
-      this.worker = (opt, cb) => this.workerFarm(JSON.stringify(opt, encode), cb);
+      this.boundWorkers = (options, cb) => this.workers(JSON.stringify(options, encode), cb);
     } else {
-      this.worker = (opt, cb) => {
+      this.boundWorkers = (options, cb) => {
         try {
-          const result = minify(opt);
-          cb(null, result);
+          cb(null, minify(options));
         } catch (errors) {
           cb(errors);
         }
       };
     }
 
-    this.worker(options, callback);
-  }
-
-  exit() {
-    if (this.workerFarm) {
-      workerFarm.end(this.workerFarm);
-    }
-  }
-
-  runTasks(tasks, callback) {
-    if (!tasks.length) {
-      callback(null, []);
-      return;
-    }
-
     let toRun = tasks.length;
     const results = [];
     const step = (index, data) => {
       toRun -= 1;
       results[index] = data;
+
       if (!toRun) {
         callback(null, results);
       }
     };
 
     tasks.forEach((task, index) => {
-      const cacheIdentifier = `${versions.uglify}|${versions.plugin}|${task.input}`;
       const enqueue = () => {
-        this.worker(task, (error, data) => {
+        this.boundWorkers(task, (error, data) => {
           const result = error ? { error } : data;
           const done = () => step(index, result);
-          if (this.cache && !result.error) {
-            put(this.cache, task.cacheKey, data, cacheIdentifier).then(done, done);
+
+          if (this.cacheDir && !result.error) {
+            cacache.put(this.cacheDir, task.cacheKey, JSON.stringify(data)).then(done, done);
```

Comment:
Does it make sense to move the cache `put` operation into the worker process?

Reply:
@sokra no sense